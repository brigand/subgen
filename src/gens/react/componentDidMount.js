// @flow
import getIndent from '../../utils/getIndent';
import jsGetLandmarks from '../utils/jsGetLandmarks';

// TODO: object type annotation
const componentDidMount = (code: string) : string => {
  const lines = code.split(/\r?\n/g);
  const indent = getIndent(code);
  const { methods, classes } = jsGetLandmarks(code);

  const firstClass = classes[0];

  if (!firstClass) throw new Error(`No class found.`);

  const ctorMethod = methods.find(({ name }) => name === `constructor`);
  const cdmMethod = methods.find(({ name }) => name === `componentDidMount`);

  if (cdmMethod) {
    return code;
  }

  const newLines = [
    `${indent}componentDidMount() {`,
    `${indent}}`,
  ];

  if (ctorMethod) {
    lines.splice(ctorMethod.end + 1, 0, ``, ...newLines);
  } else {
    lines.splice(firstClass.start + 1, 0, ...newLines);
  }

  return lines.join(`\n`);
};

export default componentDidMount;
