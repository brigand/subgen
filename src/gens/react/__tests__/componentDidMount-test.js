import outdent from 'outdent';
import componentDidMount from '../componentDidMount';

it(`works without constructor`, () => {
  const input = outdent`
  class C extends React.Component {
    foo() {

    }
    render() {

    }
  }
  `;

  const out = componentDidMount(input);

  expect(out.indexOf(`componentDidMount`)).not.toBe(-1);
  expect(out).toMatchSnapshot();
});

it(`works with constructor`, () => {
  const input = outdent`
  class C extends React.Component {
    foo() {

    }
    constructor() {

    }
    render() {

    }
  }
  `;

  const out = componentDidMount(input);

  expect(out.indexOf(`componentDidMount`)).not.toBe(-1);
  expect(out).toMatchSnapshot();
});
