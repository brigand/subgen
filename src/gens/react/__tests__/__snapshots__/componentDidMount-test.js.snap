// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works with constructor 1`] = `
"class C extends React.Component {
  foo() {

  }
  constructor() {

  }

  componentDidMount() {
  }
  render() {

  }
}"
`;

exports[`works without constructor 1`] = `
"class C extends React.Component {
  componentDidMount() {
  }
  foo() {

  }
  render() {

  }
}"
`;
